src/librustc_mir/lints.rs

```diff
@@ -0,0 +1,130 @@
+        // closures can't recur, so they don't matter.
+        return;
+    }
+
+    // Walk through this function (say `f`) looking to see if
+    // every possible path references itself, i.e. the function is
+    // called recursively unconditionally. This is done by trying
+    // to find a path from the entry node to the exit node that
+    // *doesn't* call `f` by traversing from the entry while
+    // pretending that calls of `f` are sinks (i.e. ignoring any
+    // exit edges from them).
+    //
+    // NB. this has an edge case with non-returning statements,
+    // like `loop {}` or `panic!()`: control flow never reaches
+    // the exit node through these, so one can have a function
+    // that never actually calls itself but is still picked up by
+    // this lint:
+    //
+    //     fn f(cond: bool) {
+    //         if !cond { panic!() } // could come from `assert!(cond)`
+    //         f(false)
+    //     }
+    //
+    // In general, functions of that form may be able to call
+    // itself a finite number of times and then diverge. The lint
+    // considers this to be an error for two reasons, (a) it is
+    // easier to implement, and (b) it seems rare to actually want
+    // to have behaviour like the above, rather than
+    // e.g. accidentally recursing after an assert.
+
+    let basic_blocks = mir.basic_blocks();
+    let mut work_queue = vec![mir::START_BLOCK];
+    let mut reached_exit_without_self_call = false;
+    let mut self_call_locations = vec![];
+    let mut visited = BitSet::new_empty(basic_blocks.len());
+
+    let param_env = tcx.param_env(def_id);
+    let caller_substs = Substs::identity_for_item(tcx, def_id);
+
+    while let Some(bb_id) = work_queue.pop() {
+        if visited.contains(bb_id) {
+            //already done
+            continue;
+        }
+
+        visited.insert(bb_id);
+
+        let block = &basic_blocks[bb_id];
+
+        if let Some(ref terminator) = block.terminator {
+            match terminator.kind {
+                TerminatorKind::Call { ref func, .. } => {
+                    let func_ty = func.ty(mir, tcx);
+
+                    if let TyKind::FnDef(fn_def_id, substs) = func_ty.sty {
+                        let (call_fn_id, call_substs) =
+                            if let Some(instance) = Instance::resolve(tcx,
+                                                                        param_env,
+                                                                        fn_def_id,
+                                                                        substs) {
+                                (instance.def_id(), instance.substs)
+                            } else {
+                                (fn_def_id, substs)
+                            };
+
+                        //is this a self call?
+                        let is_self_call =
+                            if let Some(Def::Method(_)) = tcx.describe_def(def_id) {
```

Comment:
Do you need this `describe_def` check? I'd suspect that if you have issues with the substs, you'd *rather* normalize them somehow. Polymorphic recursion wouldn't compile *anyway*.

Reply:
Just doing the substs check missed a few cases which the old lint caught. For example:

```
fn test<T: Copy>() {
  test::<usize>();
}
```

The current lint catches this. With the additional check for "are we in a trait method", all the existing tests pass.

Reply:
Ah, that is unfortunate, since it's not polymorphic recursion, and doesn't involve traits. I don't like `describe_def` though - can you use some functionality on `tcx` to check whether the function is a trait-associated item? Inherent methods, or methods in impls, need to behave like ordinary functions, it's only trait impls that care about dispatch.

Furthermore, *only* the part of the substs that corresponds to the trait, is relevant here, e.g. if the method has generics, they should probably be ignored.